src/pal/src/misc/miscpalapi.cpp

```diff
@@ -298,3 +301,153 @@ PAL_Random(
+static pthread_mutex_t *g_OpenSslLocks;
+
+#define CRYPTO_LOCK 1
+typedef void(*locking_function)(int mode, int n, char* file, int line);
+typedef int(*CRYPTO_num_locks)(void);
+typedef void(*CRYPTO_set_locking_callback)(locking_function callback);
+
+void LockingCallback(int mode, int n, char* file, int line)
+{
+    int result;
+    if (mode & CRYPTO_LOCK)
+    {
+        result = pthread_mutex_lock(&g_OpenSslLocks[n]);
+    }
+    else
+    {
+        result = pthread_mutex_unlock(&g_OpenSslLocks[n]);
+    }
+
+    if (result != 0)
+    {
+        ASSERT("LockingCallback(%d, %d, %s, %d) failed with error %d \n",
+            mode, n, file, line, result);
+    }
+}
+
+/*++
+Function:
+  EnsureOpenSslInitialized
+
+  Used by cryptographic libraries in CoreFX to initialize
+  threading support in OpenSSL.
+
+  --*/
+
+DWORD
+PALAPI
+EnsureOpenSslInitialized()
+{
+    DWORD dwRet = 0;
+    int numLocks;
+    CRYPTO_num_locks numLocksFunc;
+    CRYPTO_set_locking_callback setCallbackFunc;
+    int locksInitialized = 0;
+
+    PERF_ENTRY(EnsureOpenSslInitialized);
+    ENTRY("EnsureOpenSslInitialized()\n");
+
+    pthread_mutex_lock(&g_OpenSslInitLock);
+    if (g_OpenSslLocks != NULL)
+    {
+        // Already initialized; nothing more to do.
+        goto done;
+    }
+
+    // Open the libcrypto library
+    g_OpenSslLib = dlopen(libcryptoName, RTLD_NOW);
+    if (g_OpenSslLib == NULL)
+    {
+        ASSERT("Unable to load OpenSSL with dlerror \"%s\" \n", dlerror());
+        dwRet = 1;
+        goto done;
+    }
+
+    // Get the functions we need from OpenSSL
+    numLocksFunc = (CRYPTO_num_locks) dlsym(g_OpenSslLib, "CRYPTO_num_locks");
+    setCallbackFunc = (CRYPTO_set_locking_callback) dlsym(g_OpenSslLib, "CRYPTO_set_locking_callback");
+    if (numLocksFunc == NULL || setCallbackFunc == NULL)
+    {
+        ASSERT("Unable to find CRYPTO_num_locks or CRYPTO_set_locking_callback\n");
+        dwRet = 2;
+        goto done;
+    }
+
+    // Determine how many locks are needed
+    numLocks = numLocksFunc();
+    if (numLocks <= 0)
+    {
+        ASSERT("CRYPTO_num_locks returned invalid value: %d\n", numLocks);
+        dwRet = 3;
+        goto done;
+    }
+
+    // Create the locks array
+    g_OpenSslLocks = (pthread_mutex_t*) PAL_malloc(sizeof(pthread_mutex_t) * numLocks);
+    if (g_OpenSslLocks == NULL)
+    {
+        ASSERT("PAL_malloc failed\n");
+        dwRet = 4;
+        goto done;
+    }
+
+    // Initialize each of the locks
+    for (locksInitialized = 0; locksInitialized < numLocks; locksInitialized++)
+    {
+        if (pthread_mutex_init(&g_OpenSslLocks[locksInitialized], NULL) != 0)
+        {
+            ASSERT("pthread_mutex_init failed\n");
+            dwRet = 5;
+            goto done;
+        }
+    }
+
+    // Initialize the callback
+    setCallbackFunc((locking_function) LockingCallback);
+
+done:
+    pthread_mutex_unlock(&g_OpenSslInitLock);
```

Comment:
I don't know how likely it would be, particularly given that I don't know that anything here would retry into success, but it seems like you could have thread 1 fail, release the mutex, thread 2 grabs the mutex and initializes g_OpenSslLocks, then thread 1 destroyes it.  Wouldn't it be safer to release the mutex after the potential cleanup?